rfcs/0191-worker-manager-launch-configs.md

```diff
@@ -0,0 +1,122 @@
+# RFC 0191 - Worker Manager launch configurations
+* Comments: [#0191](https://github.com/taskcluster/taskcluster-rfcs/pull/191)
+* Proposed by: @lotas
+
+## Summary
+
+This proposal outlines enhancements to the worker-manager service that would address the following:
+
+* make worker pools launch configurations first-class citizens in the worker-manager service
+* balanced distribution of workers across multiple launch configurations (regions)
+* dynamic adjustment of launch configuration likelihood to be provisioned based on **health metrics**
+
+**Health metrics** - is the combination of the number of fail-to-success provisioning attempts ratio, the number of workers currently running.
+
+## Motivation
+
+The current worker-manager lacks the ability to deal with broken configurations, monitor the health of launch configurations and the availability of cloud resources.
+
+This often leads to the creation of workers that are stuck in provisioning or fail to become operational, wasting resources and reducing system efficiency.
+
+By introducing health tracking and dynamic adjustment mechanisms, we aim to optimize the provisioning process and enhance system reliability.
+
+## Details
+
+New entity will be introduced to worker-manager - **Launch Configuration**.
+Each worker pool will have zero or more **launch configurations** associated with it.
+
+Each **launch configuration** will be assigned a unique ID (hash of its properties), stored in the database and will be immutable (except for the status flag).
+
+Generated ID will include the following properties:
+
+* top-level config properties (`minCapacity`, `maxCapacity`, `scalingRatio`, `lifecyle`)
```

Comment:
My idea was that since I want to keep launch configs immutable between updates, this hash would tell that something changed.. 
For example, if wp update changes `lifecycle.registrationTimeout`, but launch configs stay the same, I want to "invalidate" (or reset) existing launch configs, and treat them as new. So the stats would be counted from zero again, as lifecycle parameters + image itself might behave differently.. i.e. worker needs more time to bootstrap

But you are right, maybe capacity and scaling ratio don't affect this much, so we can remove them

Reply:
I think invalidating all existing configs would be a rather disruptive event -- for example if a region is down, information about that status will be lost.

It seems like all of these top-level values might be changed frequently when trying to address a capacity issue, so _not_ invalidating all knowledge of lauch-config health when that occurs seems smart.

Reply:
Also specifically for the `lifecycle.registrationTimeout` example, if the launch config hash was the same but the health all of a sudden started to deteriorate, then that seems like a clue that the problem lies outside the launch config (which all of these things are).

Are you worried that issues might be detected too slowly if the launch config has a long history of being healthy? Maybe the answer to that could be to make the algorithm very aggressive in the face of new failures. E.g, three failures in a row and health is already at 0 (just an example)

Reply:
Agreed, probably not worth taking top-level options into account to preserve more information for the launch configs and keep it simple.

Reply:
My only worry was that there could be a correlation between `lifecycle.registrationTimeout` and worker "health". If timeout is too small (for a given image) then it will not have time to launch. But this is probably over-thinking already :) Happy to keep it as dumb as possible initially